drv/gimlet-hf-server/src/main.rs

```diff
@@ -147,11 +151,96 @@ fn main() -> ! {
     };
 
+    server.ensure_persistent_data_is_redundant().unwrap(); // TODO: log this?
+
     loop {
         idol_runtime::dispatch(&mut buffer, &mut server);
     }
 }
 
+////////////////////////////////////////////////////////////////////////////////
+
+/// Represents persistent data that is both stored on the host flash and used to
+/// configure host boot.
+///
+/// We reserve sector 0 (i.e. the lowest 64 KiB) of both host flash ICs for
+/// Hubris data.  On both host flashes, we tile sector 0 with this 40-byte
+/// struct, placed at 128-byte intervals (in case it needs to grow in the
+/// future).
+///
+/// The current value of persistent data is the instance of `RawPersistentData`
+/// with a valid checksum and the highest `epoch` across both flash ICs.
+///
+/// When writing new data, we increment the epoch and write to both ICs, one by
+/// one.  This ensures robustness in case of power loss.
+#[derive(Copy, Clone, AsBytes, FromBytes)]
+#[repr(C)]
+struct RawPersistentData {
+    /// Reserved field, because this is placed at address 0, which PSP firmware
+    /// may look at under certain circumstances.
+    amd_reserved_must_be_all_ones: u64,
+
+    /// Must always be 0x1dea_bcde.
+    oxide_magic: u32,
+
+    /// Must always be 1 (for now)
+    header_version: u32,
+
+    /// Monotonically increasing counter
+    epoch: u64,
+
+    /// Equivalent to `host_sp_messages::HostStartupOptions`
+    host_startup_options: u64,
+
+    /// Either 0 or 1; directly translatable to `gimlet_hf_api::HfDevSelect`
+    dev_select: u32,
+
+    /// CRC-32 over the rest of the data using the iSCSI polynomial
+    checksum: u32,
+}
+
+impl RawPersistentData {
+    fn new(data: HfPersistentData, epoch: u64) -> Self {
+        let mut out = Self {
+            amd_reserved_must_be_all_ones: u64::MAX,
+            oxide_magic: HF_PERSISTENT_DATA_MAGIC,
+            header_version: HF_PERSISTENT_DATA_HEADER_VERSION,
+            epoch,
+            host_startup_options: data.startup_options,
+            dev_select: data.dev_select as u32,
+            checksum: 0,
+        };
+        out.checksum = out.expected_checksum();
+        assert!(out.is_valid());
+        out
+    }
+
+    fn expected_checksum(&self) -> u32 {
+        let c = crc::Crc::<u32>::new(&crc::CRC_32_ISCSI);
+        let mut c = c.digest();
+        // We do a CRC32 of everything except the checksum, which is positioned
+        // at the end of the struct and is a `u32`
+        let size = core::mem::size_of::<RawPersistentData>()
+            - core::mem::size_of::<u32>();
+        c.update(&self.as_bytes()[..size]);
+        c.finalize()
+    }
+
+    fn is_valid(&self) -> bool {
+        self.amd_reserved_must_be_all_ones == u64::MAX
+            && self.oxide_magic == HF_PERSISTENT_DATA_MAGIC
+            && self.header_version == HF_PERSISTENT_DATA_HEADER_VERSION
+            && self.dev_select <= 1
+            && self.checksum == self.expected_checksum()
```

Comment:
Do we also want to check that all set bits of `self.host_startup_options` are valid `HostStartupOptions`? I suspect _not_, because that feels like some fragile over-coupling, but am not completely sure so wanted to ask.

Reply:
I considered it, but agree that it would be a little too much coupling.